extension/opampextension/opamp_agent.go

```diff
@@ -0,0 +1,244 @@
+	"os"
+	"runtime"
+
+	"github.com/oklog/ulid/v2"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	semconv "go.opentelemetry.io/collector/semconv/v1.18.0"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opamp-go/client"
+	"github.com/open-telemetry/opamp-go/client/types"
+	"github.com/open-telemetry/opamp-go/protobufs"
+)
+
+// TODO: Replace with https://github.com/open-telemetry/opentelemetry-collector/issues/6596
+const localConfig = `
+exporters:
+  otlp:
+    endpoint: localhost:1111
+receivers:
+  otlp:
+    protocols:
+      grpc: {}
+      http: {}
+service:
+  pipelines:
+    traces:
+      receivers: [otlp]
+      processors: []
+      exporters: [otlp]
+`
+
+type opampAgent struct {
+	cfg    *Config
+	logger *zap.Logger
+
+	agentType    string
+	agentVersion string
+
+	instanceId ulid.ULID
+
+	effectiveConfig string
+
+	agentDescription *protobufs.AgentDescription
+
+	opampClient client.OpAMPClient
+
+	remoteConfigStatus *protobufs.RemoteConfigStatus
+}
+
+func (o *opampAgent) Start(_ context.Context, _ component.Host) error {
+	o.opampClient = client.NewWebSocket(o.logger.Sugar())
+
+	header := http.Header{}
+	for k, v := range o.cfg.Headers {
+		header.Set(k, string(v))
+	}
+
+	tls := &tls.Config{
+		ServerName:         o.cfg.TLSSetting.ServerName,
+		InsecureSkipVerify: o.cfg.TLSSetting.InsecureSkipVerify,
+	}
+
+	settings := types.StartSettings{
+		Header:         header,
+		TLSConfig:      tls,
+		OpAMPServerURL: o.cfg.Endpoint,
+		InstanceUid:    o.instanceId.String(),
+		Callbacks: types.CallbacksStruct{
+			OnConnectFunc: func() {
+				o.logger.Debug("Connected to the OpAMP server")
+			},
+			OnConnectFailedFunc: func(err error) {
+				o.logger.Error("Failed to connect to the OpAMP server", zap.Error(err))
+			},
+			OnErrorFunc: func(err *protobufs.ServerErrorResponse) {
+				o.logger.Error("OpAMP server returned an error response", zap.String("message", err.ErrorMessage))
+			},
+			GetEffectiveConfigFunc: func(ctx context.Context) (*protobufs.EffectiveConfig, error) {
+				return o.composeEffectiveConfig(), nil
+			},
+			OnMessageFunc: o.onMessage,
+		},
+		// Include ReportsEffectiveConfig once the extension has access to the
+		// collector's effective configuration.
+		Capabilities: protobufs.AgentCapabilities_AgentCapabilities_ReportsStatus,
+	}
+
+	if err := o.createAgentDescription(); err != nil {
+		return err
+	}
+
+	if err := o.opampClient.SetAgentDescription(o.agentDescription); err != nil {
+		return err
+	}
+
+	o.logger.Debug("Starting OpAMP client...")
+
+	if err := o.opampClient.Start(context.Background(), settings); err != nil {
+		return err
+	}
+
+	o.logger.Debug("OpAMP client started")
+
+	return nil
+}
+
+func (o *opampAgent) Shutdown(ctx context.Context) error {
+	o.logger.Debug("OpAMP agent shutting down...")
+	if o.opampClient != nil {
+		o.logger.Debug("Stopping OpAMP client...")
+		err := o.opampClient.Stop(ctx)
+		return err
+	}
+	return nil
```

Comment:
nit: this can be flattened by returning from the simple case early
```suggestion
	if o.opampClient == nil {
		return nil
	}
	o.logger.Debug("Stopping OpAMP client...")
	return o.opampClient.Stop(ctx)	
```